packages/flutter_tools/lib/src/commands/update_packages.dart

```diff
@@ -1231,26 +1224,25 @@ class PubspecDependency extends PubspecLine {
   final String version; // the version string if found, or blank.
   final bool isTransitive; // whether the suffix matched kTransitiveMagicString
   final String sourcePath; // the filename of the pubspec.yaml file, for error messages
-  bool isDevDependency; // Whether this dependency is under the `dev dependencies` section.
+  bool? isDevDependency; // Whether this dependency is under the `dev dependencies` section.
```

Comment:
Is the null bool meaningful, or can this be
```suggestion
  bool isDevDependency = false; // Whether this dependency is under the `dev dependencies` section.
```

Reply:
null means we haven't set it yet. This shouldn't be read before it's set, so I think `late` would make more sense.

Reply:
made late

Reply:
I've come around to @jonahwilliams's way of thinking about `late` (not `late final`): with few exceptions it should be avoided, except in tests.  It usually points to some state needing to be known before the object is initialized, but we're setting it later, leaving the object in an invalid state.  For example,
```dart
          final PubspecDependency? dependency = PubspecDependency.parse(line, filename: filename);
          if (dependency != null) { // We got one!
            // Track whether or not this a dev dependency.
            dependency.isDevDependency = seenDev;
```
could instead be:
```dart
          final PubspecDependency? dependency = PubspecDependency.parse(line, filename: filename, isDevDependency: seenDev);
          if (dependency != null) { // We got one!
```
```dart
class PubspecDependency extends PubspecLine {
  PubspecDependency(
    String line,
    this.name,
    this.suffix, {
    required this.isTransitive,
    required DependencyKind kind,
    required this.version,
    required this.sourcePath,
    required this.isDevDependency,
  }) : _kind = kind,
       super(line);

  static PubspecDependency? parse(String line, { required String filename, required bool isDevDependency }) {
...
    return PubspecDependency(line, package, suffix, isTransitive: isTransitive, version: version, kind: stripped.isEmpty ? DependencyKind.unknown : DependencyKind.normal, sourcePath: filename, isDevDependency: isDevDependency);

...
  final bool isDevDependency;
```

Reply:
> I've come around to @jonahwilliams's way of thinking about `late` (not `late final`): with few exceptions it should be avoided, except in tests. It usually points to some state needing to be known before the object is initialized, but we're setting it later, leaving the object in an invalid state. For example,
> 
> ```dart
>           final PubspecDependency? dependency = PubspecDependency.parse(line, filename: filename);
>           if (dependency != null) { // We got one!
>             // Track whether or not this a dev dependency.
>             dependency.isDevDependency = seenDev;
> ```
> 
> could instead be:
> 
> ```dart
>           final PubspecDependency? dependency = PubspecDependency.parse(line, filename: filename, isDevDependency: seenDev);
>           if (dependency != null) { // We got one!
> ```
> 
> ```dart
> class PubspecDependency extends PubspecLine {
>   PubspecDependency(
>     String line,
>     this.name,
>     this.suffix, {
>     required this.isTransitive,
>     required DependencyKind kind,
>     required this.version,
>     required this.sourcePath,
>     required this.isDevDependency,
>   }) : _kind = kind,
>        super(line);
> 
>   static PubspecDependency? parse(String line, { required String filename, required bool isDevDependency }) {
> ...
>     return PubspecDependency(line, package, suffix, isTransitive: isTransitive, version: version, kind: stripped.isEmpty ? DependencyKind.unknown : DependencyKind.normal, sourcePath: filename, isDevDependency: isDevDependency);
> 
> ...
>   final bool isDevDependency;
> ```

I agree with you that it is a code smell and that doing the parsing in a factory would be better. I would prefer to do refactors in a separate PR, lest I break something in them and it gets reverted. How about if I file a tracking issue to clean up this file?

Reply:
Sure.  Or don't and we can just fix it if it someday breaks.  This is an internal-ish development feature we don't need to spend too much time on it.

Reply:
Also the `isDevDependency` one could be done without much refactoring by just adding it to the constructor as in my code snippet. `_lockIsOverride` is trickier.